### ISSUE TITLE
[Bug]DAPM status always off with nocodec mode

### ISSUE THREAD

**Issue by RDharageswari** [thesofproject/linux#1987]
Firmware, Topology and the kernel is from TGL_v0.4 Release.
 aplay -Dhw:0,0 /dev/urandom -c2 -f s32_le -r 48000
and arecord -Dhw:0,0 -c4 -fS32_LE try.wav -r 48000 in parallel windows.

localhost /sys/kernel/debug/asoc/sof-nocodec/0000:00:1f.3/dapm # cat * | grep Off
BUF1.0: Off  in 1 out 0
BUF1.1: Off  in 1 out 0
BUF1.2: Off  in 0 out 0
BUF2.0: Off  in 0 out 1
BUF2.1: Off  in 0 out 1
DSM1.0: Off  in 1 out 0
Echo Reference Capture 0: Off  in 0 out 1
MUXDEMUX2.0: Off  in 0 out 1
PCM0C: Off  in 0 out 1
PCM0P: Off  in 1 out 0
PIPELINE.1.SSP0.OUT: Off  in 0 out 0
PIPELINE.2.SSP0.IN: Off  in 0 out 0
SSP0.IN: Off  in 0 out 1
SSP0.OUT: Off  in 1 out 0
Smart Amplifier Playback 0: Off  in 1 out 0
Off
localhost /sys/kernel/debug/asoc/sof-nocodec/0000:00:1f.3/dapm #
None of the dapm widgets are "On"

**Comment by keyonjie** [comment]
good finding, thanks @RDharageswari

**Comment by keyonjie** [comment]
Hi @RDharageswari this is an common issue for our nocodec I2S support.

Hi @ClarexZhou @xiulipan can you help check if we can reproduce the issue on other nocodec machine?

Hi @plbossart @ranj063 @kv2019i if we think it is a real issue, we might need to fix it for the nocodec machine driver.

**Comment by ClarexZhou** [comment]
Tested on APL up2 no codec, error also seen.
```

root@sh-apl-up2-nocodec-01:/sys/kernel/debug/asoc/sof-nocodec/0000:00:0e.0/dapm# cat * | grep Off
Off
BUF1.0: Off  in 1 out 0
BUF1.1: Off  in 1 out 0
BUF11.0: Off  in 0 out 0
BUF11.1: Off  in 0 out 0
BUF1.2: Off  in 1 out 0
BUF12.0: Off  in 0 out 0
BUF12.1: Off  in 0 out 0
BUF1.3: Off  in 1 out 0
BUF13.0: Off  in 0 out 0
BUF14.0: Off  in 0 out 0
BUF14.1: Off  in 0 out 0
BUF14.2: Off  in 0 out 0
BUF2.0: Off  in 0 out 1
BUF2.1: Off  in 0 out 1
BUF3.0: Off  in 0 out 0
BUF3.1: Off  in 0 out 0
BUF4.0: Off  in 0 out 0
BUF4.1: Off  in 0 out 0
BUF7.0: Off  in 0 out 0
BUF7.1: Off  in 0 out 0
BUF8.0: Off  in 0 out 0
BUF8.1: Off  in 0 out 0
DMIC0.IN: Off  in 0 out 0
Low Latency Playback 0: Off  in 1 out 0
Media Playback 7: Off  in 0 out 0
MIXER1.0: Off  in 1 out 0
Passthrough Capture 0: Off  in 0 out 1
Passthrough Capture 1: Off  in 0 out 0
Passthrough Capture 3: Off  in 0 out 0
Passthrough Capture 5: Off  in 0 out 0
Passthrough Capture 6: Off  in 0 out 0
Passthrough Playback 1: Off  in 0 out 0
Passthrough Playback 3: Off  in 0 out 0
Passthrough Playback 5: Off  in 0 out 0
PCM0C: Off  in 0 out 1
PCM0P: Off  in 1 out 0
PCM1C: Off  in 0 out 0
PCM1P: Off  in 0 out 0
PCM3C: Off  in 0 out 0
PCM3P: Off  in 0 out 0
PCM5C: Off  in 0 out 0
PCM5P: Off  in 0 out 0
PCM6C: Off  in 0 out 0
PCM7P: Off  in 0 out 0
PGA1.0: Off  in 1 out 0
PGA1.1: Off  in 1 out 0
PGA11.0: Off  in 0 out 0
PGA12.0: Off  in 0 out 0
PGA14.0: Off  in 0 out 0
PGA2.0: Off  in 0 out 1
PGA3.0: Off  in 0 out 0
PGA4.0: Off  in 0 out 0
PGA7.0: Off  in 0 out 0
PGA8.0: Off  in 0 out 0
PIPELINE.11.SSP5.OUT: Off  in 0 out 0
PIPELINE.12.SSP5.IN: Off  in 0 out 0
PIPELINE.13.DMIC0.IN: Off  in 0 out 0
PIPELINE.14.SSP0.OUT: Off  in 0 out 0
PIPELINE.1.SSP0.OUT: Off  in 0 out 0
PIPELINE.2.SSP0.IN: Off  in 0 out 0
PIPELINE.3.SSP1.OUT: Off  in 0 out 0
PIPELINE.4.SSP1.IN: Off  in 0 out 0
PIPELINE.7.SSP3.OUT: Off  in 0 out 0
PIPELINE.8.SSP3.IN: Off  in 0 out 0
SRC14.0: Off  in 0 out 0
SSP0.IN: Off  in 0 out 1
SSP0.OUT: Off  in 1 out 0
SSP1.IN: Off  in 0 out 0
SSP1.OUT: Off  in 0 out 0
SSP3.IN: Off  in 0 out 0
SSP3.OUT: Off  in 0 out 0
SSP5.IN: Off  in 0 out 0
SSP5.OUT: Off  in 0 out 0
```
[dmesg.txt](https://github.com/thesofproject/linux/files/4448369/dmesg.txt)
[sof-logger_full.txt](https://github.com/thesofproject/linux/files/4448370/sof-logger_full.txt)

**Comment by keyonjie** [comment]
Thanks @ClarexZhou .

So this is a common nocodec issue.

**Comment by kv2019i** [comment]
@RDharageswari @keyonjie Is there some practical problems caused by this? Obviously would be good to fix this, but as nocodec is a test vehicle in the end, not sure of the priority of this issue unless this is causing actual problems to something.

**Comment by keyonjie** [comment]
@kv2019i haven't seen any real problems with this, so I agree with setting this to P2 or P3.

**Comment by plbossart** [comment]
Could this be because the sof_nocodec_card does not have any dapm widgets?

````c
static struct snd_soc_card sof_nocodec_card = {
	.name = "nocodec", /* the sof- prefix is added by the core */
	.owner = THIS_MODULE
};
````

@ranj063 does this ring a bell?

**Comment by ranj063** [comment]
> Could this be because the sof_nocodec_card does not have any dapm widgets?
> 
> ```c
> static struct snd_soc_card sof_nocodec_card = {
> 	.name = "nocodec", /* the sof- prefix is added by the core */
> 	.owner = THIS_MODULE
> };
> ```
> 
> @ranj063 does this ring a bell?

@plbossart sorry missed this earlier. My guess is it is because we dont have a codec dai in the path. I can investigate a bit more and update

**Comment by ranj063** [comment]
@plbossart so this is mainly because of the fact that the endpoint widgets are missing in the dummy codec component driver. 
This results in 0 audio paths when a PCM is started. Unfortunately, the fix for this isnt as simple as adding those EP widgets in the codec component driver either because when probing components during card registration, the dummy component is ignored. https://github.com/thesofproject/linux/blob/8ce8c4f3c93571e4e6e6dcd798c018f829ede6fb/sound/soc/soc-core.c#L1162

So, if we want to fix this, it is going to take a bit of surgery in the core. Let me know!

**Comment by ranj063** [comment]
It looks like this commit was purposely added to avoid probing the dummy component and avoid creating widgets for the dummy DAI. 6e78108bda78adbb2d4ef55ec60a388aba975797

**Comment by plbossart** [comment]
I am still struggling to understand the severity of this sighting. Is DAPM widgets are off, does this mean DAPM is also disabled de-facto with dummy-dais?

**Comment by ranj063** [comment]
> I am still struggling to understand the severity of this sighting. Is DAPM widgets are off, does this mean DAPM is also disabled de-facto with dummy-dais?

@plbossart it certainly looks like DAPM is disabled for the dummy dais. There are no output/input widgets or routes in the definition of the dummy codec. I added those in the dummy codec as @bardliao  suggested and removed the patch https://github.com/thesofproject/linux/commit/6e78108bda78adbb2d4ef55ec60a388aba975797 and now I can see that the widgets are on as they're supposed to be. What I added were the DUMMYRX and DUMMY OUT widgets and the route to connect them.

![widget](https://user-images.githubusercontent.com/7766921/90045935-94e43900-dc84-11ea-8701-e3a515225e7a.png)

**Comment by ranj063** [comment]
The patch disables DAPM for a specific reason that the dummy dai's can used by multiple cards. One way I think we can fix this problem if we really care about the DAPM power status with the nocodec mode is define our own sof-dummy-codec but I dont know how it would be recevied upstream

**Comment by plbossart** [comment]
could we make that behavior configurable and opt-in, e.g. in the card we add a field that says we want to deal with the dummy-dais.

**Comment by ranj063** [comment]
> could we make that behavior configurable and opt-in, e.g. in the card we add a field that says we want to deal with the dummy-dais.

@plbossart what if multiple cards set the same field?

**Comment by plbossart** [comment]
> > could we make that behavior configurable and opt-in, e.g. in the card we add a field that says we want to deal with the dummy-dais.
> 
> @plbossart what if multiple cards set the same field?

I still don't get what the issue was with multiple cards to be honest

**Comment by ranj063** [comment]
@plbossart if we had multiple cards using the dummy dai, we'd fail here while checking if the component is bound to this card or not.
https://github.com/thesofproject/linux/blob/8ce8c4f3c93571e4e6e6dcd798c018f829ede6fb/sound/soc/soc-core.c#L1166

**Comment by ranj063** [comment]
On a side note, do you think we should add some DAPM power status related test to our test suite? It might be helpful in debugging issues with PM.

**Comment by plbossart** [comment]
> @plbossart if we had multiple cards using the dummy dai, we'd fail here while checking if the component is bound to this card or not.
> https://github.com/thesofproject/linux/blob/8ce8c4f3c93571e4e6e6dcd798c018f829ede6fb/sound/soc/soc-core.c#L1

This gives me a migraine.

**Comment by ranj063** [comment]
> > @plbossart if we had multiple cards using the dummy dai, we'd fail here while checking if the component is bound to this card or not.
> > https://github.com/thesofproject/linux/blob/8ce8c4f3c93571e4e6e6dcd798c018f829ede6fb/sound/soc/soc-core.c#L1
> 
> This gives me a migraine.

@plbossart maybe I can check with Mark as to what we can do about this scenario. I think it would be useful for us in terms of debugging PM issues with the nocodec mode to have DAPM enabled. What do you think?

**Comment by ranj063** [comment]
@plbossart after seeing Mark's response, @bardliao and I feel that this is expected behaviour with the dummy DAIs and probably left as it is. Adding a simple codec just to see the DAPM power updates doesnt feel like it is really critical for the nocodec machine driver. What do you think?

**Comment by plbossart** [comment]
@ranj063 @bardliao we have more interesting things to do, I vote for a WON'T FIX status. This is a framework problem with almost no ROI for us.

**Comment by mengdonglin** [comment]
@ranj063 it seems Mark doesn't want to enable DAPM for Dummy DAIs, but suggest introduce a "simple codec driver":  https://mailman.alsa-project.org/pipermail/alsa-devel/2020-August/172463.html

Do you think we can close this issue as won't fix and create a new feature to follow up?

**Comment by keyonjie** [comment]
This is actually still a hole for us if we can't fix it, which will make the nocodec machine driver much less useful for us, as we can't use it to cover any PM related features.

**Comment by plbossart** [comment]
> This is actually still a hole for us if we can't fix it, which will make the nocodec machine driver much less useful for us, as we can't use it to cover any PM related features.

what failure are you referring to here? this looks hypothetical to me.
Also I did mention that you guys are abusing nocodec topologies for your CI tests...

**Comment by ranj063** [comment]
> @ranj063 it seems Mark doesn't want to enable DAPM for Dummy DAIs, but suggest introduce a "simple codec driver": https://mailman.alsa-project.org/pipermail/alsa-devel/2020-August/172463.html
> 
> Do you think we can close this issue as won't fix and create a new feature to follow up?

@mengdonglin Im not sure we need to create a new feature for follow up. Im not sure what advantage adding this simple codec will bring other than being able to see the DAPM power status of the widgets.

**Comment by keyonjie** [comment]
> > This is actually still a hole for us if we can't fix it, which will make the nocodec machine driver much less useful for us, as we can't use it to cover any PM related features.
> 
> what failure are you referring to here? this looks hypothetical to me.

like check-runtime-pm-double-active.sh, check-runtime-pm-status.sh, these cases will fail with nocodec machine.

> Also I did mention that you guys are abusing nocodec topologies for your CI tests...

um, IMHO, it depends on how you define the "nocodec", if the intention of the existence of this "nocodec" is to help people verify kinds of features on platforms where the parts of endpoints are not ready (e.g. SSP codecs not connected, i915 driver is not ready, DMICs not connected, ...), then enabling/enriching these possibilities with the only need to create new validation topologies sounds charming to me, why not?

**Comment by plbossart** [comment]
> > > This is actually still a hole for us if we can't fix it, which will make the nocodec machine driver much less useful for us, as we can't use it to cover any PM related features.
> > 
> > 
> > what failure are you referring to here? this looks hypothetical to me.
> 
> like check-runtime-pm-double-active.sh, check-runtime-pm-status.sh, these cases will fail with nocodec machine.

No, they should not. pm_runtime should work regardless on DAPM being enabled or not.
 
> > Also I did mention that you guys are abusing nocodec topologies for your CI tests...
> 
> um, IMHO, it depends on how you define the "nocodec", if the intention of the existence of this "nocodec" is to help people verify kinds of features on platforms where the parts of endpoints are not ready (e.g. SSP codecs not connected, i915 driver is not ready, DMICs not connected, ...), then enabling/enriching these possibilities with the only need to create new validation topologies sounds charming to me, why not?

I think we have to be very careful. Every time we introduce test tools, we need to have a means to 'test the tester', and I really don't see this mindset so far.

The intent of the nocodec mode was that we could inject or retrieve data to verify the audio interfaces with e.g. Logic Pro or Audio Precision. With all the features you mention, there is a very limited way to *test* the results, and I worry that in a lot of cases this will give a false impression that things work, without the ability to check, or report issues that don't exist on a real platform. So yes, nice idea on paper but how does one actually test all this?

**Comment by keyonjie** [comment]
> > > > This is actually still a hole for us if we can't fix it, which will make the nocodec machine driver much less useful for us, as we can't use it to cover any PM related features.
> > > 
> > > 
> > > what failure are you referring to here? this looks hypothetical to me.
> > 
> > 
> > like check-runtime-pm-double-active.sh, check-runtime-pm-status.sh, these cases will fail with nocodec machine.
> 
> No, they should not. pm_runtime should work regardless on DAPM being enabled or not.

The DAPM widgets status will impact the pm_runtime status, if they are off where they should actually be on, then there could be error like the PCI device is in runtime suspended where it should be active, we hit this kind of issues.

> 
> > > Also I did mention that you guys are abusing nocodec topologies for your CI tests...
> > 
> > 
> > um, IMHO, it depends on how you define the "nocodec", if the intention of the existence of this "nocodec" is to help people verify kinds of features on platforms where the parts of endpoints are not ready (e.g. SSP codecs not connected, i915 driver is not ready, DMICs not connected, ...), then enabling/enriching these possibilities with the only need to create new validation topologies sounds charming to me, why not?
> 
> I think we have to be very careful. Every time we introduce test tools, we need to have a means to 'test the tester', and I really don't see this mindset so far.
> 
> The intent of the nocodec mode was that we could inject or retrieve data to verify the audio interfaces with e.g. Logic Pro or Audio Precision. With all the features you mention, there is a very limited way to _test_ the results, and I worry that in a lot of cases this will give a false impression that things work, without the ability to check, or report issues that don't exist on a real platform. So yes, nice idea on paper but how does one actually test all this?

The point for "test the tester" is good, that's what we have been doing with reporting and fixing issues to sof-test.

The active part which motivate us to implement this kind of validation is that it do have already helped us on finding issues, among them some could be difficult to emerge with normal codecs connected platforms, see:
https://github.com/thesofproject/sof/issues/3317

**Comment by keyonjie** [comment]
After discussed with @bardliao, with adding endpoint widgets to machine driver or codec driver, the widget power status is under control via DAPM, but we have to introduce an extra switch mixer to switch on/off the connection b/w DSM playback and capture streams as what the max98373 does today, and doing the switching via UCM in enable/disable sequence.

Considering this is significant effort even including ones outside of the driver, let's leave this as known issue and won't fix it until somebody has a better solution.

**Comment by keyonjie** [comment]
Close it as won't fix, we can reopen once we want to revisit it.
